### ISSUE TITLE
Ссылки на функции

### ISSUE THREAD

**Issue by tormozit** [SeiOkami/OneS#205]
Уже есть заявка "Функции как объекты" https://github.com/SeiOkami/OneS/issues/14
Но она очень сложная в восприятии и уверен в реализации.
Я же предлагаю намного более простой и потому реалистичный вариант - только ссылки на функции.
```
Функция Сложить(Число1, Число2)
	коэф = 0.1;
	Возврат (Число1 + Число2) * коэф;
КонецФункции

Процедура Тест()
	Ф = Функция Сложить;
	Сообщить(ф(1, 2));
КонецПроцедуры
```

**Comment by Hobbit-Jedi** [comment]
А как предлагается использовать <КонтекстныйОбъект>?

**Comment by tormozit** [comment]
> А как предлагается использовать <КонтекстныйОбъект>?

Если метод принадлежит модулю размножаемого объекта (например СправочникОбъект или Форма), то вызывать его можно только от объекта, которому он принадлежит. Если конструктор вызывается внутри модуля, в котором есть метод-параметр, то можно не указывать объект - платформа сама его привяжет.

**Comment by Hobbit-Jedi** [comment]
А если мы создали ссылку на метод от объекта, а потом объект вышел из области видимости (удален уже из памяти), а мы вызываем метод по ссылке? Должно вывалиться красивое исключение?

**Comment by Hobbit-Jedi** [comment]
В-общем, примеров использования этого параметра не хватает.

**Comment by tormozit** [comment]
> А если мы создали ссылку на метод от объекта, а потом объект вышел из области видимости (удален уже из памяти)

Такое невозможно, ссылка на объект в ссылке метода будет удерживать объект в памяти.

**Comment by Hobbit-Jedi** [comment]
> Удаление объекта из памяти в такой ситуации невозможно. Ссылка на объект в ссылке метода будет удерживать объект в памяти.

Ладно. Не совсем из памяти.

Например,
```
&НаКлиенте
Процедура ОбработчикНажатияКнопки()
  Ф1 = ПолучитьСсылкуНаФункциюНаСервере();
  ВызовМетода Ф1(1, 2); // Здесь содержащего функцию объекта уже нет.
КонецПроцедуры

&НаСервере
Функция ПолучитьСсылкуНаФункциюНаСервере()
  ТекущийОбъект = РеквизитФормыВЗначение("Объект");
  Результат = Новый СсылкаМетода("РассчитатьСумму1",  ТекущийОбъект);
  Возврат Результат;
КонецФункции
```

**Comment by tormozit** [comment]
Я по сути предлагаю опрятную замену существующим приемам через "Выполнить()". Исключения будут возникать аналогичные. Если объект не существует, то и вызвать его метод нельзя - будет исключение выброшено.

**Comment by Nivanchenko** [comment]
Мне больше нравиться конструкция вызова, как то вот так:
```bsl
СсылкаМетода.Выполнить(Параметры);
```
Или в идеале

```bsl
СсылкаМетода = Новый СсылкаМетода(ИмяМетода, КонтекстныйОбъект);
СсылкаМетода(Параметры);
```

**Comment by tormozit** [comment]
> СсылкаМетода.Выполнить(Параметры);

Согласен. Так не придется расширять синтаксис языка. Переделал на твой вариант.

**Comment by SeiOkami** [comment]
Думаю, что даже если на сервер добавят "ОписаниеОповещения", то этого бы хватило для большинства случаев. Объект можно передавать в качестве параметра. Да, описанное решение удобнее, но тут вопрос в сложности реализации на уровне платформы. Идеально всё же #14 , но если делать минимум, то можно просто ОписаниеОповещения на сервере позволять описывать.

**Comment by tormozit** [comment]
ОписаниеОповещения слишком неудобно, т.к. 
1. требует для всех методов делать обретки с двумя параметрами "Результат, ДопПараметры"
2. требует сразу передать все параметры в месте создания "ссылки" на метод (по сути это не ссылка на метод, а описание вызова метода с конкретными параметрами)

**Comment by SeiOkami** [comment]
Отправлено боту 28.04.2023

**Comment by tormozit** [comment]
Все же в некоторых случаях, например при подключении обработчика ожидания, будет удобно сразу в ссылке на метод указать значения параметров для вызова. Поэтому добавим опциональный 3-й параметр в конструктор аналогично объекту ОписаниеОповещения:
`Новый СсылкаМетода(<ИмяМетода>, <КонтекстныйОбъект>[, <ЗначенияПараметров>])`

**Comment by tormozit** [comment]
Еще один костыльный способ вызывать по имени функцию с 3-мя параметрами (в 3-м передаваемая структура). Работает во всех контекстах при условии доступности экспорта методов.
```
Чт = Новый ЧтениеJSON;
Чт.УстановитьСтроку("{'r':1}");
я = ПрочитатьJSON(Чт,,,, "Функ", ЭтотОбъект, Струкутура);
Функция Функ(_1,_2, Струкутура) Экспорт
	Сообщить(1);
КонецФункции
```

**Comment by tormozit** [comment]
Реализации похожих возможностей через метод Выполнить() в БСП:
- ОбщегоНазначения.ВыполнитьМетодКонфигурации()
- ОбщегоНазначения.ВыполнитьМетодОбъекта()
- ДлительныеОперации.ВыполнитьФункцию()
- ДлительныеОперации.ВыполнитьПроцедуру()

**Comment by tormozit** [comment]
Эта возможность также позволит сделать оптимальной работу пока прикладного объекта [Функтор](https://infostart.ru/1c/articles/2639512), который уже запоминает и значения параметров вместе с указателем на метод. Возможно в платформе лучше сразу делать аналогично, т.е. со значениями параметров.

**Comment by tormozit** [comment]
Осознал, что возможность указания параметров в конструкторе и хранение их между вызовами дает важные преимущества. Дополнил параметры конструктора в полном описании:
`Новый СсылкаМетода(<КонтекстныйОбъект>, <ИмяМетода> [, П1, П2, ...])`
